File: server/schemas/workflow.py
"""Schemas for Workflow CRUD."""

from __future__ import annotations

from typing import Any
from datetime import datetime

from pydantic import BaseModel


class FieldDef(BaseModel):
    name: str
    label: str
    field_type: str = "text"  # text|number|date|phone|id_card|file|select|multi_select|address|custom
    required: bool = True
    validation_rule: str | None = None  # regex or built-in rule name
    placeholder: str = ""
    options: list[dict[str, Any]] | None = None  # for select/multi_select
    # File upload config (only for field_type="file")
    file_config: dict[str, Any] | None = None  # {max_size_mb, allowed_extensions, storage_path}
    # LLM-assisted validation (semantic checks like "is this a valid address?")
    llm_validate: bool = False
    llm_validate_prompt: str | None = None


class StepCreate(BaseModel):
    name: str
    order: int
    step_type: str = "collect"
    prompt_template: str = ""
    fields: list[FieldDef] | None = None
    validation_rules: dict[str, Any] | None = None
    tool_id: str | None = None
    tool_config: dict[str, Any] | None = None
    on_failure: str = "retry"
    max_retries: int = 2
    fallback_step_id: str | None = None
    requires_human_confirm: bool = False
    risk_level: str = "info"
    next_step_rules: dict[str, Any] | None = None


class StepOut(BaseModel):
    id: str
    workflow_id: str
    name: str
    order: int
    step_type: str
    prompt_template: str
    fields: Any
    validation_rules: Any
    tool_id: str | None
    tool_config: Any
    on_failure: str
    max_retries: int
    fallback_step_id: str | None
    requires_human_confirm: bool
    risk_level: str
    next_step_rules: Any
    created_at: datetime

    model_config = {"from_attributes": True}


class WorkflowCreate(BaseModel):
    name: str
    description: str = ""
    tenant_id: str = "default"
    config: dict[str, Any] | None = None
    steps: list[StepCreate] | None = None


class WorkflowUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    config: dict[str, Any] | None = None


class WorkflowOut(BaseModel):
    id: str
    name: str
    description: str
    tenant_id: str
    version: int
    config: Any
    steps: list[StepOut] = []
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: console/src/types/workflow.ts
export interface Workflow {
  id: string;
  name: string;
  description: string;
  tenant_id: string;
  version: number;
  config: Record<string, unknown> | null;
  steps: WorkflowStep[];
  created_at: string;
  updated_at: string;
}

export interface WorkflowStep {
  id: string;
  workflow_id: string;
  name: string;
  order: number;
  step_type: StepType;
  prompt_template: string;
  fields: FieldDef[] | null;
  validation_rules: Record<string, unknown> | null;
  tool_id: string | null;
  tool_config: Record<string, unknown> | null;
  on_failure: OnFailureStrategy;
  max_retries: number;
  fallback_step_id: string | null;
  requires_human_confirm: boolean;
  risk_level: string;
  next_step_rules: NextStepRule[] | null;
  created_at: string;
}

export type StepType = 'collect' | 'validate' | 'tool_call' | 'confirm' | 'human_review' | 'complete';
export type OnFailureStrategy = 'retry' | 'skip' | 'rollback' | 'escalate';

export interface FieldDef {
  name: string;
  label: string;
  field_type: string;
  required: boolean;
  validation_rule?: string | null;
  placeholder?: string;
  options?: Array<{ label: string; value: string }> | null;
  file_config?: FileConfig | null;
  llm_validate?: boolean;
  llm_validate_prompt?: string | null;
}

export interface FileConfig {
  max_size_mb?: number;
  allowed_extensions?: string[];
  storage_path?: string;
}

export interface NextStepRule {
  condition: RuleCondition | null;
  goto_step: string;
}

export interface RuleCondition {
  field: string;
  op: 'eq' | 'ne' | 'gt' | 'lt' | 'gte' | 'lte' | 'contains' | 'regex' | 'in' | 'not_in';
  value: unknown;
}

export interface StepCreate {
  name: string;
  order: number;
  step_type?: StepType;
  prompt_template?: string;
  fields?: FieldDef[];
  validation_rules?: Record<string, unknown>;
  tool_id?: string;
  tool_config?: Record<string, unknown>;
  on_failure?: OnFailureStrategy;
  max_retries?: number;
  next_step_rules?: NextStepRule[];
  requires_human_confirm?: boolean;
  risk_level?: string;
}

export interface WorkflowCreate {
  name: string;
  description?: string;
  tenant_id?: string;
  config?: Record<string, unknown>;
  steps?: StepCreate[];
}

export type WorkflowUpdate = Partial<Pick<WorkflowCreate, 'name' | 'description' | 'config'>>;

export interface WorkflowVersion {
  id: string;
  workflow_id: string;
  version: number;
  snapshot?: Record<string, unknown>;
  published_by: string;
  created_at: string;
}
